// igniteui-angular copy/esm2022/lib/animations/easings.mjs
var EaseIn;
(function (EaseIn) {
    EaseIn[EaseIn["Quad"] = `cubic-bezier(0.550, 0.085, 0.680, 0.530)`] = "Quad";
    EaseIn[EaseIn["Cubic"] = `cubic-bezier(0.550, 0.055, 0.675, 0.190)`] = "Cubic";
    EaseIn[EaseIn["Quart"] = `cubic-bezier(0.895, 0.030, 0.685, 0.220)`] = "Quart";
    EaseIn[EaseIn["Quint"] = `cubic-bezier(0.755, 0.050, 0.855, 0.060)`] = "Quint";
    EaseIn[EaseIn["Sine"] = `cubic-bezier(0.470, 0.000, 0.745, 0.715)`] = "Sine";
    EaseIn[EaseIn["Expo"] = `cubic-bezier(0.950, 0.050, 0.795, 0.035)`] = "Expo";
    EaseIn[EaseIn["Circ"] = `cubic-bezier(0.600, 0.040, 0.980, 0.335)`] = "Circ";
    EaseIn[EaseIn["Back"] = `cubic-bezier(0.600, -0.280, 0.735, 0.045)`] = "Back";
})(EaseIn || (EaseIn = {}));
var EaseOut;
(function (EaseOut) {
    EaseOut[EaseOut["Quad"] = `cubic-bezier(0.250, 0.460, 0.450, 0.940)`] = "Quad";
    EaseOut[EaseOut["Cubic"] = `cubic-bezier(0.215, 0.610, 0.355, 1.000)`] = "Cubic";
    EaseOut[EaseOut["Quart"] = `cubic-bezier(0.165, 0.840, 0.440, 1.000)`] = "Quart";
    EaseOut[EaseOut["Quint"] = `cubic-bezier(0.230, 1.000, 0.320, 1.000)`] = "Quint";
    EaseOut[EaseOut["Sine"] = `cubic-bezier(0.390, 0.575, 0.565, 1.000)`] = "Sine";
    EaseOut[EaseOut["Expo"] = `cubic-bezier(0.190, 1.000, 0.220, 1.000)`] = "Expo";
    EaseOut[EaseOut["Circ"] = `cubic-bezier(0.075, 0.820, 0.165, 1.000)`] = "Circ";
    EaseOut[EaseOut["Back"] = `cubic-bezier(0.175, 0.885, 0.320, 1.275)`] = "Back";
})(EaseOut || (EaseOut = {}));
var EaseInOut;
(function (EaseInOut) {
    EaseInOut[EaseInOut["Quad"] = `cubic-bezier(0.455, 0.030, 0.515, 0.955)`] = "Quad";
    EaseInOut[EaseInOut["Cubic"] = `cubic-bezier(0.645, 0.045, 0.355, 1.000)`] = "Cubic";
    EaseInOut[EaseInOut["Quart"] = `cubic-bezier(0.770, 0.000, 0.175, 1.000)`] = "Quart";
    EaseInOut[EaseInOut["Quint"] = `cubic-bezier(0.860, 0.000, 0.070, 1.000)`] = "Quint";
    EaseInOut[EaseInOut["Sine"] = `cubic-bezier(0.445, 0.050, 0.550, 0.950)`] = "Sine";
    EaseInOut[EaseInOut["Expo"] = `cubic-bezier(1.000, 0.000, 0.000, 1.000)`] = "Expo";
    EaseInOut[EaseInOut["Circ"] = `cubic-bezier(0.785, 0.135, 0.150, 0.860)`] = "Circ";
    EaseInOut[EaseInOut["Back"] = `cubic-bezier(0.680, -0.550, 0.265, 1.550)`] = "Back";
})(EaseInOut || (EaseInOut = {}));
export { EaseIn, EaseOut, EaseInOut };
